File: server/common/src/main/java/br/edu/ufcg/virtus/common/service/SearchService.java
package br.edu.ufcg.virtus.common.service;

import java.io.Serializable;

import br.edu.ufcg.virtus.common.dto.PageListDTO;
import br.edu.ufcg.virtus.common.dto.SearchFilterDTO;
import br.edu.ufcg.virtus.common.exception.BusinessException;
import br.edu.ufcg.virtus.common.model.Model;
import br.edu.ufcg.virtus.common.repository.SearchBaseRepository;

/**
 * The 'SearchService' provides the common API for all services that do search operations with models.
 *
 * All search model services MUST extend this class.
 *
 * @author Virtus
 *
 * @param <M>
 *            Model type.
 * @param <T>
 *            ID type.
 */
public abstract class SearchService<M extends Model<T>, T extends Serializable> extends BaseService {

    /**
     * Gets the search Repository.
     *
     * @return Search Repository.
     */
    protected abstract SearchBaseRepository<M, T> getRepository();

    /**
     * Searches the model with the filter.
     *
     * @param filter
     *            Filter.
     * @return Instances founded.
     */
    public PageListDTO search(SearchFilterDTO filter) {
        return this.getRepository().search(filter);
    }

    /**
     * Gets one model instance by ID.
     *
     * @param id
     *            ID.
     * @return Instance founded.
     * @throws BusinessException
     */
    public M getOne(T id) throws BusinessException {
        return this.getRepository().findById(id).get();
    }
}

File: client/common/projects/common/src/lib/services/search.service.ts
import { Injectable } from '@angular/core';

import { BaseService } from './base.service';
import { HttpParams } from '@angular/common/http';

/**
 * The 'SearchService' class provides the common API and operations to
 * retrieve, find, filter and list items.
 *
 * @extends BaseService
 */
@Injectable()
export class SearchService extends BaseService {
    /**
     * Constructor.
     */
    constructor() {
        super();
    }

    /**
     * Gets all items.
     *
     * @param url - The url of backend service
     * @returns The result of the search method.
     */
    getAll(url) {
        return this.search(url);
    }

    /**
     * Searches the items where the filters matches.
     *
     * @param url - The url of backend service
     * @param pagination - The pagination object
     * @param filters - The filters to search
     * @returns The result values of a paginated and filtered search at the
     * backend services
     */
    search(url, pagination = {}, filters = []): any {
        const params = new HttpParams().set(
            'filter',
            JSON.stringify(Object.assign(pagination, { filters }))
        );

        return this.get(url, params);
    }

    /**
     * Gets one item by its ID.
     *
     * @param url - The url of backend service
     * @param id - The id of the iten
     * @returns The item by its id
     */
    getOne(url, id) {
        return this.get(url + '/' + id);
    }

}
